src/zarr/core/metadata/v3.py

```diff
@@ -297,71 +332,99 @@ def update_attributes(self, attributes: dict[str, JSON]) -> Self:
+COMPLEX_DTYPE = Literal["complex64", "complex128"]
 COMPLEX = np.complex64 | np.complex128
+STRING_DTYPE = Literal["string"]
+STRING = np.str_
+# BYTES_DTYPE = np.dtypes.BytesDType
+BYTES_DTYPE = Literal["bytes"]
+BYTES = np.bytes_
+
+ALL_DTYPES = INTEGER_DTYPE | FLOAT_DTYPE | COMPLEX_DTYPE | STRING_DTYPE | BYTES_DTYPE
 
 
 @overload
 def parse_fill_value(
-    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool | None,
+    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool,
     dtype: BOOL_DTYPE,
 ) -> BOOL: ...
 
 
 @overload
 def parse_fill_value(
-    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool | None,
+    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool,
     dtype: INTEGER_DTYPE,
 ) -> INTEGER: ...
 
 
 @overload
 def parse_fill_value(
-    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool | None,
+    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool,
     dtype: FLOAT_DTYPE,
 ) -> FLOAT: ...
 
 
 @overload
 def parse_fill_value(
-    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool | None,
+    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool,
     dtype: COMPLEX_DTYPE,
 ) -> COMPLEX: ...
 
 
 @overload
 def parse_fill_value(
-    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool | None,
-    dtype: np.dtype[Any],
-) -> Any:
-    # This dtype[Any] is unfortunately necessary right now.
-    # See https://github.com/zarr-developers/zarr-python/issues/2131#issuecomment-2318010899
-    # for more details, but `dtype` here (which comes from `parse_dtype`)
-    # is np.dtype[Any].
-    #
-    # If you want the specialized types rather than Any, you need to use `np.dtypes.<dtype>`
-    # rather than np.dtypes(<type>)
-    ...
+    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool,
+    dtype: STRING_DTYPE,
+) -> STRING: ...
+
+
+@overload
+def parse_fill_value(
+    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool,
+    dtype: BYTES_DTYPE,
+) -> BYTES: ...
+
+
+# @overload
+# def parse_fill_value(
+#     fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool | None,
+#     dtype: np.dtype[Any],
+# ) -> Any:
+# # This dtype[Any] is unfortunately necessary right now.
+# # See https://github.com/zarr-developers/zarr-python/issues/2131#issuecomment-2318010899
+# # for more details, but `dtype` here (which comes from `parse_dtype`)
+# # is np.dtype[Any].
+# #
+# # If you want the specialized types rather than Any, you need to use `np.dtypes.<dtype>`
+# # rather than np.dtypes(<type>)
+# ...
 
 
 def parse_fill_value(
-    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool | None,
-    dtype: BOOL_DTYPE | INTEGER_DTYPE | FLOAT_DTYPE | COMPLEX_DTYPE | np.dtype[Any],
+    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool,
+    dtype_value: ALL_DTYPES,
```

Comment:
Mypy is not happy here.

```
src/zarr/core/metadata/v3.py:220: error: No overload variant of "parse_fill_value" matches argument types "Any", "Literal['string', 'bytes'] | Literal['int8', 'int16', 'int32', 'int64', 'uint8', 'uint16', 'uint32', 'uint64'] | Literal['float16', 'float32', 'float64'] | Literal['complex64', 'complex128']"  [call-overload]
src/zarr/core/metadata/v3.py:220: note: Possible overload variants:
src/zarr/core/metadata/v3.py:220: note:     def parse_fill_value(fill_value: complex | str | bytes | generic | Sequence[Any] | bool, dtype: Literal['bool']) -> bool
src/zarr/core/metadata/v3.py:220: note:     def parse_fill_value(fill_value: complex | str | bytes | generic | Sequence[Any] | bool, dtype: Literal['int8', 'int16', 'int32', 'int64', 'uint8', 'uint16', 'uint32', 'uint64']) -> signedinteger[_8Bit] | signedinteger[_16Bit] | signedinteger[_32Bit] | signedinteger[_64Bit] | unsignedinteger[_8Bit] | unsignedinteger[_16Bit] | unsignedinteger[_32Bit] | unsignedinteger[_64Bit]
src/zarr/core/metadata/v3.py:220: note:     def parse_fill_value(fill_value: complex | str | bytes | generic | Sequence[Any] | bool, dtype: Literal['float16', 'float32', 'float64']) -> floating[_16Bit] | floating[_32Bit] | floating[_64Bit]
src/zarr/core/metadata/v3.py:220: note:     def parse_fill_value(fill_value: complex | str | bytes | generic | Sequence[Any] | bool, dtype: Literal['complex64', 'complex128']) -> complexfloating[_32Bit, _32Bit] | complexfloating[_64Bit, _64Bit]
src/zarr/core/metadata/v3.py:220: note:     def parse_fill_value(fill_value: complex | str | bytes | generic | Sequence[Any] | bool, dtype: Literal['string']) -> str_
src/zarr/core/metadata/v3.py:220: note:     def parse_fill_value(fill_value: complex | str | bytes | generic | Sequence[Any] | bool, dtype: Literal['bytes']) -> bytes_
src/zarr/core/metadata/v3.py:425: error: Overloaded function implementation does not accept all possible arguments of signature 1  [misc]
src/zarr/core/metadata/v3.py:425: error: Overloaded function implementation does not accept all possible arguments of signature 2  [misc]
src/zarr/core/metadata/v3.py:425: error: Overloaded function implementation does not accept all possible arguments of signature 3  [misc]
src/zarr/core/metadata/v3.py:425: error: Overloaded function implementation does not accept all possible arguments of signature 4  [misc]
src/zarr/core/metadata/v3.py:425: error: Overloaded function implementation does not accept all possible arguments of signature 5  [misc]
src/zarr/core/metadata/v3.py:425: error: Overloaded function implementation does not accept all possible arguments of signature 6  [misc]
src/zarr/core/metadata/v3.py:514: error: Returning Any from function declared to return "str | bytes | generic"  [no-any-return]
```

Reply:
I think 7d9d89736245c8e547c9cda0b70683de74982dfd fixed this. Two things:

1. Some of the overloads used `dtype` and some used `dtype` and some used `dtype_value`. They need to match
2. `BOOL_DTYPE` wasn't present in `ALL_DTYPES`, and mypy's error message didn't make this clear :/ I ended up running that file through pyright which made it a bit easier to see:

```
  /Users/tom/gh/zarr-developers/zarr-python/src/zarr/core/metadata/v3.py:410:5 - error: Overloaded implementation is not consistent with signature of overload 1
    Type "(fill_value: Any, dtype: ALL_DTYPES) -> Any" is not assignable to type "(fill_value: complex | str | bytes | generic | Sequence[Any] | bool, dtype: BOOL_DTYPE) -> BOOL"
      Parameter 2: type "BOOL_DTYPE" is incompatible with type "ALL_DTYPES"
        Type "BOOL_DTYPE" is not assignable to type "ALL_DTYPES"
          "BOOL_DTYPE" is not assignable to type "Literal['int8']"
          "BOOL_DTYPE" is not assignable to type "Literal['int16']"
          "BOOL_DTYPE" is not assignable to type "Literal['int32']"
          "BOOL_DTYPE" is not assignable to type "Literal['int64']"
          "BOOL_DTYPE" is not assignable to type "Literal['uint8']"
    ... (reportInconsistentOverload)
15 errors, 0 warnings, 0 informations 
```

Reply:
Thanks Tom! Really appreciate your help here!